### ISSUE TITLE
Quick runs may not throw exceptions

### ISSUE THREAD

**Issue by lie112** [APSIMInitiative/ApsimX#7103]
It appears that a very fast simulation (sub second) can be complete before the exception handling is performed and the flow moves to halting the simulation thread. Therefore, APSIM says the run was successful even with an exception thrown and errors reported in the Messages table.
This can also happen if background checks finish the simulation quickly with errors reported.
Should we add a final check of errors presented before writing success and giving successful completion sound?

**Comment by hol430** [comment]
Where is your exception being thrown from? I just created a short simulation (runs in 0.01 seconds), and if I modify it to generate an exception, the exception is displayed as expected.

**Comment by lie112** [comment]
The exception is thrown in a model (CLEM\Resources\Finance.cs) during the CLEMValidate event so after StartOfSimulation. 
On building this example for you I found it was all working until I emptied the datastore after which the problem occurs. 
[ExceptionTest.zip](https://github.com/APSIMInitiative/ApsimX/files/8077017/ExceptionTest.zip)

**Comment by hol430** [comment]
It's possible to "run" a folder. When you do, all simulations under the folder will be run. In your example file you have a reports folder with no simulations in it, so if you run that folder, nothing will be run. I think the real problem here is that we get a message saying "Reports complete (.01 sec)" when in fact nothing has been run. In this case it would probably be better to treat the empty folder as any other non-simulation model.

Could that explain what you were seeing here? Or were you running into something else entirely?

**Comment by lie112** [comment]
These were run either in an experiment, or from the simulations or simulation models (see difference below) and it seems that the thread is closed and success reported before the errors are reported. I do not try and run empty folders and the empty report folder is unused.

I have been able to further diagnose the problem

Step 1 When first run,  the error is thrown because the validation check on the month of financial year fails. This is written to the messages as an error in CLEMValidate. 

Step 2 after correcting the issue and running the model again (right click the Simulation (TestActivityControl) component and select Run the error persists even though it has been fixed. I.e. the messages in the datastore have not been cleared

Step 3 If I run the model from right click Simulations (Top of tree) the model runs successfully as it must clear the messages.

Step 4 put the error back by setting financial year month to Not Set, Run from SimulationS and the models runs even though there's an error

Step 5 Run again from Simulation (TestActivityControl), runs ok

Step 6 Empty datastore and run from SimulationS (ok

Step 7 Run from Simulation (TestActivityControl) ok

Step 8 Close APSIM and start again. Run from Simulation and error is again reported

I can't figure out why success is written before the validation message is reported unless we restart APSIM. This can get confusing for users as the error is not up to date with the issue when resolved, or caused.







As stated above this problem seems to occur when there is no (or empty) datastore so I am wondering if the errors haven't been written before "success" is reported. I have slowly identified and fixed issues where error reporting was missed

Step 1 From exploring now, the error is thrown because the validation check on the month of financial year fails. This is written to the messages as an error in CLEMValidate. 

Step 2. If I go an correct this issue and run the model again I get the same error

I have noticed that once this error is thrown is keeps getting thrown even after fixed (month set) if running the Simulation. Then maybe running the Simulations clears the error list and the simulation runs.
